ddtrace/contrib/jinja2/patch.py

```diff
@@ -0,0 +1,80 @@
+import jinja2
+from wrapt import wrap_function_wrapper as _w
+
+from ddtrace import tracer
+
+from ...ext import http
+from ...utils.wrappers import unwrap as _u
+
+DEFAULT_TEMPLATE_NAME = "<template>"
+
+
+def patch():
+    if getattr(jinja2, "__datadog_patch", False):
+        # already patched
+        return
+    setattr(jinja2, "__datadog_patch", True)
+    _w(jinja2, "environment.Template.render", _wrap_render)
+    _w(jinja2, "environment.Template.generate", _wrap_generate)
+    _w(jinja2, "environment.Environment.compile", _wrap_compile)
+    _w(jinja2, "environment.Environment._load_template", _wrap_load_template)
+
+
+def unpatch():
+    if not getattr(jinja2, "__datadog_patch", False):
+        return
+    setattr(jinja2, "__datadog_patch", False)
+    _u(jinja2.Template, "render")
+    _u(jinja2.Template, "generate")
+    _u(jinja2.Environment, "compile")
+    _u(jinja2.Environment, "_load_template")
+
+
+def _wrap_generate(func, instance, args, kwargs):
+    template_name = instance.name or DEFAULT_TEMPLATE_NAME
+    with tracer.trace("jinja2.generate", span_type=http.TEMPLATE) as span:
+        try:
+            for s in func(*args, **kwargs):
+                yield s
+        finally:
+            span.resource = template_name
+            span.set_tag("jinja2.template_name", template_name)
+
+
+def _wrap_render(func, instance, args, kwargs):
+    template_name = instance.name or DEFAULT_TEMPLATE_NAME
+    with tracer.trace("jinja2.render", span_type=http.TEMPLATE) as span:
+        try:
+            return func(*args, **kwargs)
+        finally:
+            span.resource = template_name
+            span.set_tag("jinja2.template_name", template_name)
+
+
+def _wrap_compile(func, instance, args, kwargs):
+    if len(args) > 1:
+        template_name = args[1]
+    else:
+        template_name = kwargs.get("name", DEFAULT_TEMPLATE_NAME)
+
+    with tracer.trace("jinja2.compile", span_type=http.TEMPLATE) as span:
+        try:
+            return func(*args, **kwargs)
+        finally:
+            span.resource = template_name
+            span.set_tag("jinja2.template_name", template_name)
+
+
+def _wrap_load_template(func, instance, args, kwargs):
+    template_name = kwargs.get("name", args[0])
+    with tracer.trace("jinja2.load", span_type=http.TEMPLATE) as span:
+        template = None
+        try:
+            template = func(*args, **kwargs)
+        finally:
+            span.resource = template_name
+            span.set_tag("jinja2.template_name", template_name)
+            loader = instance.loader.__class__.__name__
+            span.set_tag("jinja2.loader", loader)
+            span.set_tag("jinja2.template_path", template.filename)
```

Comment:
`template` could be `None` here if a `jinja2.exceptions.TemplateNotFound` was raised.

Also, do all templates have a `filename`?